### PR TITLE
Fuse row-wise sharded linear matmul to increase perf.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78598
* #78477
* __->__ #78449

Instead of looping through and performing a matmul separately, we can
just perform a single matmul to ensure we launch a single cuda kernel for this
operation.

Differential Revision: [D36743354](https://our.internmc.facebook.com/intern/diff/D36743354/)

Differential Revision: [D36743354](https://our.internmc.facebook.com/intern/diff/D36743354)

### CODE REVIEW

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/linear.py [hunk 1]
```diff
@@ -357,8 +353,9 @@ def forward(ctx, world_size, start_pos, chunk_size, weight, pg, bias):
     @staticmethod
     def backward(ctx, grad_output):
         results = []
+        grad_output_cloned = grad_output.clone()
         for idx in range(ctx.world_size):
-            results.append(grad_output.clone())
+            results.append(grad_output_cloned)
```
Comment (vadimkantorov) [inline torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/linear.py:358]: Could this be reformulated as `results = [grad_output.clone()] * ctx.world_size`?

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/linear.py [hunk 2]
```diff
@@ -323,21 +323,17 @@ def _handle_row_wise_sharding_sharded_tensor(
...
         raise NotImplementedError(
             "The case when the input does not come from col-wise sharded "
             "linear is not supported for row-wise sharded linear."
         )
 
-    for tensor in torch.tensor_split(local_shard, world_size):
-        results.append(
-            tensor.matmul(local_shard_t) + _BiasTensorPartial.apply(world_size, bias)
-        )
+    result = local_shard.matmul(local_shard_t) + _BiasTensorPartial.apply(world_size, bias)
```
Comment (aazzolini) [inline torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/linear.py:333]: Would addmm allow to fuse the bias addition?

**Review by aazzolini — APPROVED** [review]

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/linear.py [hunk 3]
```diff
@@ -323,21 +323,17 @@ def _handle_row_wise_sharding_sharded_tensor(
     Returns:
         A :class:`_PartialTensor` object which stores the partial local result.
     """
-    results = []
-    local_shard = input.local_shards()[0].tensor
+    local_input = input.local_shards()[0].tensor
```
Comment (wanchaol) [inline torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/linear.py:326]: nit: input.local_tensor() to trigger the check of single local shard.

**Review by wanchaol — APPROVED** [review]
Looks great!

**Pull request closed without merging**